engine/src/main/java/org/pentaho/di/www/CarteServlet.java

```diff
@@ -0,0 +1,174 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2019 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.www;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.pentaho.di.core.exception.KettlePluginException;
+import org.pentaho.di.core.logging.LogChannel;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.plugins.CartePluginType;
+import org.pentaho.di.core.plugins.PluginInterface;
+import org.pentaho.di.core.plugins.PluginRegistry;
+import org.pentaho.di.core.plugins.PluginTypeListener;
+
+public class CarteServlet extends HttpServlet {
+
+  private static final long serialVersionUID = 2434694833497859776L;
+
+  public static final String STRING_CARTE_SERVLET = "Carte Servlet";
+
+  private Map<String, CartePluginInterface> cartePluginRegistry;
+
+  private final LogChannelInterface log;
+  private List<SlaveServerDetection> detections;
+
+  public CarteServlet() {
+    this.log = new LogChannel( STRING_CARTE_SERVLET );
+  }
+
+  public void doPost( HttpServletRequest req, HttpServletResponse resp ) throws ServletException, IOException {
+    doGet( req, resp );
+  }
+
+  public void doGet( HttpServletRequest req, HttpServletResponse resp ) throws ServletException, IOException {
+    String servletPath = req.getPathInfo();
+    if ( servletPath != null && servletPath.endsWith( "/" ) ) {
+      servletPath = servletPath.substring( 0, servletPath.length() - 1 );
+    }
+    CartePluginInterface plugin = cartePluginRegistry.get( servletPath );
+    if ( plugin != null ) {
+      try {
+        plugin.doGet( req, resp );
+      } catch ( ServletException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)